File: AocCsharp/Day07/Day07.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using AocCsharp.Utils;

namespace AocCsharp.Day07;

public static class Day07
{
    public static void Run()
    {
        var testInputPath = Path.Combine("..", "inputs", "day07_test.txt");
        var realInputPath = Path.Combine("..", "inputs", "day07.txt");

        TestRunner.RunSolution("Part 1", Part1, testInputPath, realInputPath, expectedTestResult: 21);
        TestRunner.RunSolution("Part 2", Part2, testInputPath, realInputPath, expectedTestResult: 40);
    }

    public static long Part1(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0) return 0;

        var grid = lines.Select(l => l.ToCharArray()).ToArray();
        int height = grid.Length;
        int width = grid[0].Length;

        // Find starting position 'S'
        int startRow = -1, startCol = -1;
        for (int row = 0; row < height; row++)
        {
            for (int col = 0; col < width; col++)
            {
                if (grid[row][col] == 'S')
                {
                    startRow = row;
                    startCol = col;
                    break;
                }
            }
            if (startRow != -1) break;
        }

        // Simulate beams moving downward
        // Each beam is represented by its column position at each row
        // We'll track active beams as they move down row by row
        var currentBeams = new HashSet<int> { startCol };
        int splitCount = 0;

        for (int row = startRow + 1; row < height; row++)
        {
            var nextBeams = new HashSet<int>();

            foreach (int col in currentBeams)
            {
                char cell = grid[row][col];

                if (cell == '^')
                {
                    // Splitter encountered - beam stops, creates two new beams
                    splitCount++;

                    // Add beams to the left and right
                    if (col - 1 >= 0)
                    {
                        nextBeams.Add(col - 1);
                    }
                    if (col + 1 < width)
                    {
                        nextBeams.Add(col + 1);
                    }
                }
                else
                {
                    // Empty space - beam continues downward
                    nextBeams.Add(col);
                }
            }

            currentBeams = nextBeams;

            // If no beams left, we're done
            if (currentBeams.Count == 0)
            {
                break;
            }
        }

        return splitCount;
    }

    public static long Part2(string input)
    {
        var lines = input.Replace("\r", "").Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0) return 0;

        var grid = lines.Select(l => l.ToCharArray()).ToArray();
        int height = grid.Length;
        int width = grid[0].Length;

        // Find starting position 'S'
        int startRow = -1, startCol = -1;
        for (int row = 0; row < height; row++)
        {
            for (int col = 0; col < width; col++)
            {
                if (grid[row][col] == 'S')
                {
                    startRow = row;
                    startCol = col;
                    break;
                }
            }
            if (startRow != -1) break;
        }

        // For Part 2, we track the number of distinct timelines/paths
        // Each timeline is counted by tracking how many paths lead to each position
        var currentPaths = new Dictionary<int, long> { { startCol, 1 } };

        for (int row = startRow + 1; row < height; row++)
        {
            var nextPaths = new Dictionary<int, long>();

            foreach (var kvp in currentPaths)
            {
                int col = kvp.Key;
                long pathCount = kvp.Value;
                char cell = grid[row][col];

                if (cell == '^')
                {
                    // Splitter - particle takes both paths (quantum splitting)
                    // Each existing path splits into two timelines
                    if (col - 1 >= 0)
                    {
                        if (!nextPaths.ContainsKey(col - 1))
                            nextPaths[col - 1] = 0;
                        nextPaths[col - 1] += pathCount;
                    }
                    if (col + 1 < width)
                    {
                        if (!nextPaths.ContainsKey(col + 1))
                            nextPaths[col + 1] = 0;
                        nextPaths[col + 1] += pathCount;
                    }
                }
                else
                {
                    // Empty space - particle continues downward
                    if (!nextPaths.ContainsKey(col))
                        nextPaths[col] = 0;
                    nextPaths[col] += pathCount;
                }
            }

            currentPaths = nextPaths;

            // If no paths left, we're done
            if (currentPaths.Count == 0)
            {
                break;
            }
        }

        // Sum all timelines that reach the bottom
        return currentPaths.Values.Sum();
    }
}

File: AocC/day07.c
#include "day07.h"
#include "runner.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

void day07_run(void) {
    const char* test_path = "../inputs/day07_test.txt";
    const char* real_path = "../inputs/day07.txt";

    run_solution("Part 1", day07_part1, test_path, real_path, 21);
    run_solution("Part 2", day07_part2, test_path, real_path, 40);
}

int64_t day07_part1(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    // Parse lines
    char* lines[1000];
    int height = 0;

    char* line = strtok(data, "\r\n");
    while (line && height < 1000) {
        lines[height++] = line;
        line = strtok(NULL, "\r\n");
    }

    if (height == 0) {
        free(data);
        return 0;
    }

    int width = (int)strlen(lines[0]);

    // Find starting position 'S'
    int start_row = -1, start_col = -1;
    for (int row = 0; row < height; row++) {
        for (int col = 0; col < (int)strlen(lines[row]); col++) {
            if (lines[row][col] == 'S') {
                start_row = row;
                start_col = col;
                break;
            }
        }
        if (start_row != -1) break;
    }

    if (start_row == -1) {
        free(data);
        return 0;
    }

    // Track beams using a set (simple array for small widths)
    bool current_beams[1000] = {false};
    current_beams[start_col] = true;

    int64_t split_count = 0;

    for (int row = start_row + 1; row < height; row++) {
        bool next_beams[1000] = {false};
        int line_len = (int)strlen(lines[row]);

        for (int col = 0; col < width; col++) {
            if (!current_beams[col]) continue;
            if (col >= line_len) continue;

            char cell = lines[row][col];

            if (cell == '^') {
                split_count++;
                if (col - 1 >= 0) next_beams[col - 1] = true;
                if (col + 1 < width) next_beams[col + 1] = true;
            } else {
                next_beams[col] = true;
            }
        }

        memcpy(current_beams, next_beams, sizeof(current_beams));

        // Check if any beams left
        bool has_beams = false;
        for (int i = 0; i < width; i++) {
            if (current_beams[i]) {
                has_beams = true;
                break;
            }
        }
        if (!has_beams) break;
    }

    free(data);
    return split_count;
}

int64_t day07_part2(const char* input) {
    char* data = strdup(input);
    if (!data) return 0;

    // Parse lines
    char* lines[1000];
    int height = 0;

    char* line = strtok(data, "\r\n");
    while (line && height < 1000) {
        lines[height++] = line;
        line = strtok(NULL, "\r\n");
    }

    if (height == 0) {
        free(data);
        return 0;
    }

    int width = (int)strlen(lines[0]);

    // Find starting position 'S'
    int start_row = -1, start_col = -1;
    for (int row = 0; row < height; row++) {
        for (int col = 0; col < (int)strlen(lines[row]); col++) {
            if (lines[row][col] == 'S') {
                start_row = row;
                start_col = col;
                break;
            }
        }
        if (start_row != -1) break;
    }

    if (start_row == -1) {
        free(data);
        return 0;
    }

    // Track path counts per column
    int64_t current_paths[1000] = {0};
    current_paths[start_col] = 1;

    for (int row = start_row + 1; row < height; row++) {
        int64_t next_paths[1000] = {0};
        int line_len = (int)strlen(lines[row]);

        for (int col = 0; col < width; col++) {
            if (current_paths[col] == 0) continue;
            if (col >= line_len) continue;

            char cell = lines[row][col];

            if (cell == '^') {
                if (col - 1 >= 0) next_paths[col - 1] += current_paths[col];
                if (col + 1 < width) next_paths[col + 1] += current_paths[col];
            } else {
                next_paths[col] += current_paths[col];
            }
        }

        memcpy(current_paths, next_paths, sizeof(current_paths));

        // Check if any paths left
        bool has_paths = false;
        for (int i = 0; i < width; i++) {
            if (current_paths[i] > 0) {
                has_paths = true;
                break;
            }
        }
        if (!has_paths) break;
    }

    // Sum all paths at bottom
    int64_t total = 0;
    for (int i = 0; i < width; i++) {
        total += current_paths[i];
    }

    free(data);
    return total;
}
